cc-wrapper-test: support cross compilers

NickCao opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Run compiled check programs under an emulator, in order to make cc-wrapper tests work with cross compilers.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] riscv64-linux (cross)
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

NickCao committed:
cc-wrapper-test: support cross compilers

siraben commented:
Result of `nixpkgs-review pr 163117` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

Mindavi commented:
Emulator is an empty string when not cross-compiling? Or does it always use qemu now?

NickCao commented:
> Emulator is an empty string when not cross-compiling? Or does it always use qemu now?

It would be sh if I remembered correctly.

Mindavi commented on pkgs/test/cc-wrapper/default.nix:
```diff
@@ -17,36 +18,37 @@ in stdenv.mkDerivation {
-      ./cc-static
-      ${lib.optionalString (stdenv.cc.isGNU && lib.versionAtLeast (lib.getVersion stdenv.cc.name) "8.0.0") ''
+      ${emulator} ./cc-static
+      ${lib.optionalString (stdenv.cc.isGNU && (stdenv.hostPlatform.isx86 || stdenv.hostPlatform.isAarch64)
```
Why are platforms restricted here?

NickCao replied:
I can't recall why I did that, maybe it was failing on riscv.

Mindavi commented:
Also, this seems to be broken on Darwin if I understand what @siraben posted. Can we prevent that?

NickCao commented:
Removed the platform restrictions and did a rebase, however `cc-wrapper-gcc` is now failing on `aarch64-linux`, complaining `error while loading shared libraries: libstdc++.so.6: cannot open shared object file: No such file or directory`, which, is in the rpath of the binary in question.

Mindavi commented:
I also found that the sanitizer tests are broken on Darwin. A comment in the file says that too, but they seem to be enabled anyway... https://hydra.nixos.org/build/199311100/nixlog/2

NickCao committed:
cc-wrapper-test: do not test sanitizers on darwin

NickCao commented:
Seems that sanitizers does not work under qemu-user either.

NickCao committed:
cc-wrapper-test: do not test sanitizers when cross compiling

Mindavi reviewed (approved):
I don't know about that aarch64 error, is there anything that can be done about it? I think the diff now looks good but would be great if we didn't regress anything here.

NickCao commented:
After disabling sanitizer when cross compiling, the tests now passes on aarch64, I would assume it to be some weird interaction between our cc-wrapper and qemu-user.

Mindavi commented:
@ofborg build tests.cc-wrapper-gcc tests.cc-wrapper-clang

NickCao commented:
What's wrong with our clang's address sanitizer? It's failing everywhere.

Mindavi commented:
I was wondering the same. It might've been broken already though, lemme check

NickCao commented:
Found a gcc issue, which is indeed fixed, however clang is still somehow affected, https://bugzilla.redhat.com/show_bug.cgi?id=1950244

NickCao committed:
cc-wrapper-test: add workaround for asan allocation error

Miminal program fail with address sanitizer error failed to allocate 0x0 (0) bytes of SetAlternateSignalStack
https://bugzilla.redhat.com/show_bug.cgi?id=1950244

Mindavi commented:
This does not looks great: https://hydra.nixos.org/eval/1786622?filter=cc-wrapper&compare=1786593&full=#tabs-still-fail

Mindavi commented:
Thanks for going down the rabbit hole with me. Hopefully this helps a bit :)

Mindavi commented:
I guess I forgot to press the button, thanks!
